sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -55,9 +59,88 @@ class StreamExecution(
   private val sources =
     logicalPlan.collect { case s: StreamingRelation => s.source }
 
-  // Start the execution at the current offsets stored in the sink. (i.e. avoid reprocessing data
-  // that we have already processed).
-  {
+  /** When false, signals to the microBatchThread that it should stop running. */
+  @volatile
+  private var state: State = INITIALIZED
+
+  @volatile
+  private[sql] var lastExecution: QueryExecution = null
+
+  @volatile
+  private[sql] var streamDeathCause: ContinuousQueryException = null
+
+  /** The thread that runs the micro-batches of this stream. */
+  private[sql] val microBatchThread = new Thread(s"stream execution thread for $name") {
+    override def run(): Unit = { runBatches() }
+  }
+
+  /** Whether the query is currently active or not */
+  override def isActive: Boolean = state == ACTIVE
+
+  /** Returns current status of all the sources. */
+  override def sourceStatuses: Array[SourceStatus] = {
+    sources.map(s => new SourceStatus(s.toString, streamProgress.get(s))).toArray
+  }
+
+  /** Returns current status of the sink. */
+  override def sinkStatus: SinkStatus = new SinkStatus(sink.toString, sink.currentOffset)
+
+  /** Returns the [[ContinuousQueryException]] if the query was terminated by an exception. */
+  override def exception: Option[ContinuousQueryException] = Option(streamDeathCause)
+
+  /**
+   * Starts the execution. This returns only after the thread has started and [[QueryStarted]] event
+   * has been posted to all the listeners.
+   */
+  private[sql] def start(): Unit = {
```

Comment:
@marmbrus I moved the starting of the thread to a separate method because this method has to wait for the thread to start, and i dont think its a good idea to do something blocking like that in the constructor of a class. I think its cleaner this way.

For the same reason, I moved the code to initialize offsets from sink out of the constructor, as `sink.currentOffsets` can be time consuming, and can run user code.